modin/core/dataframe/pandas/metadata/index.py

```diff
@@ -23,15 +25,93 @@ class ModinIndex:
 
     Parameters
     ----------
-    value : sequence or callable
+    value : sequence, PandasDataframe or callable() -> (pandas.Index, list of ints)
+        If a sequence passed this will be considered as the index values.
+        If a ``PandasDataframe`` passed then it will be used to lazily extract indices
+        when required, note that the `axis` parameter must be passed in this case.
+        If a callable passed then it's expected to return a pandas Index and a list of
+        partition lengths along the index axis.
+    axis : int, optional
+        Specifies an axis the object represents, serves as an optional hint. This parameter
+        must be passed in case value is a ``PandasDataframe``.
     """
 
-    def __init__(self, value):
+    def __init__(self, value, axis=None):
+        from modin.core.dataframe.pandas.dataframe.dataframe import PandasDataframe
+
+        self._is_default_callable = False
+        self._axis = axis
+
         if callable(value):
             self._value = value
+        elif isinstance(value, PandasDataframe):
+            assert axis is not None
+            self._value = self._get_default_callable(value, axis)
+            self._is_default_callable = True
         else:
             self._value = ensure_index(value)
+
         self._lengths_cache = None
+        # index/lengths ID's for faster comparison between other ModinIndex objects,
+        # these should be propagated to the copies of the index
+        self._index_id = uuid.uuid4()
+        self._lengths_id = uuid.uuid4()
+
+    @staticmethod
+    def _get_default_callable(dataframe_obj, axis):
+        """
+        Build a callable extracting index labels and partitions lengths for the specified axis.
+
+        Parameters
+        ----------
+        dataframe_obj : PandasDataframe
+        axis : int
+            0 - extract indices, 1 - extract columns.
+
+        Returns
+        -------
+        callable() -> tuple(pandas.Index, list[ints])
+        """
+        # HACK: for an unknown reason, the 'lambda' approach seems to trigger some strange
+        # race conditions in HDK on certain versions of python, causing the tests to fail
+        # (python 3.9.* and 3.10.* are the versions where we saw the problem). That's
+        # really strange, but practically the same code that uses 'functools.partial'
+        # instead of a lambda works absolutely fine.
+        # return lambda: dataframe_obj._compute_axis_labels_and_lengths(axis)
```

Comment:
Shouldn't this be explored more deeper?

Reply:
> Shouldn't this be explored more deeper?

for the record, I of course tried building different versions of HDK from their repo (including the latest one), but the problem still occurred on specifically python 3.9.* and python 3.10.*

Reply:
Thanks for the explanation! Still, I would recommend to file a reproducer if possible and investigate it when we have time.